core/src/impl/Kokkos_SharedAlloc.hpp

```diff
@@ -264,13 +264,34 @@ class SharedAllocationRecordCommon : public SharedAllocationRecord<void, void> {
   /**\brief  Reallocate tracked memory in the space */
   static void deallocate_tracked(void* arg_alloc_ptr);
   /**\brief  Deallocate tracked memory in the space */
+  template <class ExecutionSpace = typename MemorySpace::execution_space>
   static void* reallocate_tracked(void* arg_alloc_ptr, size_t arg_alloc_size);
   static auto get_record(void* alloc_ptr) -> derived_t*;
   std::string get_label() const override;
   static void print_records(std::ostream& s, MemorySpace const&,
                             bool detail = false);
 };
 
+template <class MemorySpace>
+template <class ExecutionSpace>
+void* SharedAllocationRecordCommon<MemorySpace>::reallocate_tracked(
+    void* arg_alloc_ptr, size_t arg_alloc_size) {
+  derived_t* const r_old = derived_t::get_record(arg_alloc_ptr);
+  derived_t* const r_new =
+      allocate(r_old->m_space, r_old->get_label(), arg_alloc_size);
+
+  Kokkos::Impl::DeepCopy<MemorySpace, MemorySpace>(
+      ExecutionSpace{}, r_new->data(), r_old->data(),
+      std::min(r_old->size(), r_new->size()));
+  Kokkos::fence(std::string("SharedAllocationRecord<") + MemorySpace::name() +
+                ", void>::reallocate_tracked(): fence after copying data");
+
+  record_base_t::increment(r_new);
+  record_base_t::decrement(r_old);
+
+  return r_new->data();
+}
```

Comment:
The problem here was that we don't have a full definition of the execution space instance available when instantiating the class. Templating delays the instantiation until the `reallocate_tracked` member is actually needed and its only use is in `kokkos_realloc`. We could try to just inline it and get rid of this member function.

Reply:
Something like
```diff
diff --git a/core/src/Kokkos_Core.hpp b/core/src/Kokkos_Core.hpp
index 9588d289a..f6675d4aa 100644
--- a/core/src/Kokkos_Core.hpp
+++ b/core/src/Kokkos_Core.hpp
@@ -178,6 +178,21 @@ inline void* kokkos_realloc(void* arg_alloc, const size_t arg_alloc_size) {
   using MemorySpace = typename Space::memory_space;
   return Impl::SharedAllocationRecord<MemorySpace>::reallocate_tracked(
       arg_alloc, arg_alloc_size);
+
+  auto* const r_old =
+      Impl::SharedAllocationRecord<MemorySpace>::get_record(arg_alloc);
+  auto* const r_new = Impl::SharedAllocationRecord<MemorySpace>::allocate(
+      r_old->get_space(), r_old->get_label(), arg_alloc_size);
+
+  Kokkos::Impl::DeepCopy<MemorySpace, MemorySpace>(
+      r_new->data(), r_old->data(), std::min(r_old->size(), r_new->size()));
+  Kokkos::fence(std::string("SharedAllocationRecord<") + MemorySpace::name() +
+                ", void>::reallocate_tracked(): fence after copying data");
+
+  Impl::SharedAllocationRecord<void, void>::increment(r_new);
+  Impl::SharedAllocationRecord<void, void>::decrement(r_old);
+
+  return r_new->data();
 }
 
 }  // namespace Kokkos
diff --git a/core/src/impl/Kokkos_SharedAlloc.hpp b/core/src/impl/Kokkos_SharedAlloc.hpp
index e0c5b0d7a..ddfa1bf92 100644
--- a/core/src/impl/Kokkos_SharedAlloc.hpp
+++ b/core/src/impl/Kokkos_SharedAlloc.hpp
@@ -267,6 +267,7 @@ class SharedAllocationRecordCommon : public SharedAllocationRecord<void, void> {
   static void* reallocate_tracked(void* arg_alloc_ptr, size_t arg_alloc_size);
   static auto get_record(void* alloc_ptr) -> derived_t*;
   std::string get_label() const override;
+  const MemorySpace& get_space() { return m_space; }
   static void print_records(std::ostream& s, MemorySpace const&,
                             bool detail = false);
 };
@@ -331,6 +332,7 @@ class HostInaccessibleSharedAllocationRecordCommon
                             bool detail = false);
   static auto get_record(void* alloc_ptr) -> derived_t*;
   std::string get_label() const override;
+  const MemorySpace& get_space() { return m_space; }
 };
 
 #ifdef KOKKOS_ENABLE_DEBUG
```

Reply:
Do you think this should be done in this PR?

Reply:
I could argue both ways. On the one hand, there are not that many other changes. On the other hand, the changes are somewhat orthogonal to the goal of this pull request. In the end, we need a good description anyway.

Reply:
If we do this one first, we have tp justify introducing the additional dummy template argument.

Reply:
At the very least, add a comment explaining the additional template argument here.

Reply:
Ok, I'll add comments.